cmd/kubeadm/app/cmd/phases/addons.go

```diff
@@ -17,186 +17,127 @@ limitations under the License.
-func NewCmdAddon() *cobra.Command {
-	cmd := &cobra.Command{
-		Use:     "addon",
-		Aliases: []string{"addons"},
+func NewAddonPhase() workflow.Phase {
+	return workflow.Phase{
+		Name:     "addon",
+		//TODO: add aliases to workflow.Phase
+		//Aliases: []string{"addons"},
 		Short:   "Installs required addons for passing Conformance tests",
 		Long:    cmdutil.MacroCommandLongDescription,
+		Phases:  getAddonSubPhases(),
 	}
-
-	cmd.AddCommand(getAddonsSubCommands()...)
-	return cmd
 }
 
-// EnsureAllAddons installs all addons to a Kubernetes cluster
-func EnsureAllAddons(cfg *kubeadmapi.InitConfiguration, client clientset.Interface) error {
+func getData(c workflow.RunData) (*kubeadmapi.InitConfiguration, clientset.Interface, error) {
+	data, ok := c.(addonData)
+	if !ok {
+		return nil, nil, errors.New("addon phase invoked with an invalid data struct")
+	}
+
+	cfg := data.Cfg()
+
+	client, err := data.Client()
+	if err != nil {
+		return nil, nil, err
+	}
 
-	addonActions := []func(cfg *kubeadmapi.InitConfiguration, client clientset.Interface) error{
-		dnsaddon.EnsureDNSAddon,
-		proxyaddon.EnsureProxyAddon,
+	return cfg, client, err
+}
+
+// EnsureCoreDNSAddon installs CoreDNS addon to a Kubernetes cluster
+func EnsureCoreDNSAddon(c workflow.RunData) error {
+	cfg, client, err := getData(c)
+	if err != nil {
+		return nil
 	}
+	err = dnsaddon.EnsureDNSAddon(cfg, client)
+	return err
+}
 
-	glog.V(1).Infoln("[addon] installing all addons")
-	for _, action := range addonActions {
-		err := action(cfg, client)
-		if err != nil {
-			return err
-		}
+// EnsureKubeProxyAddon installs KubeProxy addon to a Kubernetes cluster
+func EnsureKubeProxyAddon(c workflow.RunData) error {
+	cfg, client, err := getData(c)
+	if err != nil {
+		return nil
 	}
+	err = proxyaddon.EnsureProxyAddon(cfg, client)
+	return err
+}
 
-	return nil
+// EnsureAllAddons installs all addons to a Kubernetes cluster
+func EnsureAllAddons(c workflow.RunData) error {
+	err := EnsureCoreDNSAddon(c)
+	if err != nil {
+		return err
+	}
+	err = EnsureKubeProxyAddon(c)
+	return err
 }
 
-// getAddonsSubCommands returns sub commands for addons phase
-func getAddonsSubCommands() []*cobra.Command {
-	cfg := &kubeadmapiv1beta1.InitConfiguration{}
-	// Default values for the cobra help text
-	kubeadmscheme.Scheme.Default(cfg)
-
-	var cfgPath, featureGatesString string
-	var subCmds []*cobra.Command
-	kubeConfigFile := kubeadmconstants.GetAdminKubeConfigPath()
-
-	subCmdProperties := []struct {
-		use      string
-		short    string
-		long     string
-		examples string
-		cmdFunc  func(cfg *kubeadmapi.InitConfiguration, client clientset.Interface) error
-	}{
+// getAddonSubPhases returns sub phases for addons phase
+func getAddonSubPhases() []workflow.Phase {
+	return []workflow.Phase{
 		{
-			use:      "all",
-			short:    "Installs all addons to a Kubernetes cluster",
-			long:     allAddonsLongDesc,
-			examples: allAddonsExample,
-			cmdFunc:  EnsureAllAddons,
+			Name:     "all",
```

Comment:
"all" is not needed because the phase runner will run all subphases unless a sub phase is specified.

Reply:
nice!